File: src-tauri/src/geo.rs
//! 공통 지리 유틸리티 (Haversine, Bearing, Destination Point)

/// 지구 평균 반경 (m)
pub const EARTH_RADIUS_M: f64 = 6_371_000.0;
/// 지구 평균 반경 (km)
pub const EARTH_RADIUS_KM: f64 = 6_371.0;

/// Haversine 거리 (km) — clamp 가드 포함
pub fn haversine_km(lat1: f64, lon1: f64, lat2: f64, lon2: f64) -> f64 {
    let dlat = (lat2 - lat1).to_radians();
    let dlon = (lon2 - lon1).to_radians();
    let a = (dlat / 2.0).sin().powi(2)
        + lat1.to_radians().cos() * lat2.to_radians().cos() * (dlon / 2.0).sin().powi(2);
    EARTH_RADIUS_KM * 2.0 * a.clamp(0.0, 1.0).sqrt().asin()
}

/// Haversine 거리 (m)
pub fn haversine_m(lat1: f64, lon1: f64, lat2: f64, lon2: f64) -> f64 {
    haversine_km(lat1, lon1, lat2, lon2) * 1000.0
}

/// 두 지점 간 방위각 (0~360°, true north, clockwise)
pub fn bearing_deg(lat1: f64, lon1: f64, lat2: f64, lon2: f64) -> f64 {
    let lat1 = lat1.to_radians();
    let lat2 = lat2.to_radians();
    let dlon = (lon2 - lon1).to_radians();
    let y = dlon.sin() * lat2.cos();
    let x = lat1.cos() * lat2.sin() - lat1.sin() * lat2.cos() * dlon.cos();
    (y.atan2(x).to_degrees() + 360.0) % 360.0
}

/// 시작점에서 방위/거리 이동한 좌표 (거리: m)
pub fn destination_point_m(lat: f64, lon: f64, bearing: f64, distance_m: f64) -> (f64, f64) {
    let lat1 = lat.to_radians();
    let lon1 = lon.to_radians();
    let brg = bearing.to_radians();
    let d_r = distance_m / EARTH_RADIUS_M;
    let lat2 = (lat1.sin() * d_r.cos() + lat1.cos() * d_r.sin() * brg.cos()).asin();
    let lon2 = lon1
        + (brg.sin() * d_r.sin() * lat1.cos()).atan2(d_r.cos() - lat1.sin() * lat2.sin());
    (lat2.to_degrees(), lon2.to_degrees())
}

/// 시작점에서 방위/거리 이동한 좌표 (거리: km)
pub fn destination_point_km(lat: f64, lon: f64, bearing: f64, distance_km: f64) -> (f64, f64) {
    destination_point_m(lat, lon, bearing, distance_km * 1000.0)
}

File: src/utils/geo.ts
/** 공통 GIS 유틸리티 (Haversine 거리, 방위각, 방위+거리 통합) */

const R_EARTH = 6371; // km
const DEG2RAD = Math.PI / 180;

/** Haversine 대원 거리 (km) */
export function haversineKm(lat1: number, lon1: number, lat2: number, lon2: number): number {
  const dLat = (lat2 - lat1) * DEG2RAD;
  const dLon = (lon2 - lon1) * DEG2RAD;
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(lat1 * DEG2RAD) * Math.cos(lat2 * DEG2RAD) * Math.sin(dLon / 2) ** 2;
  return R_EARTH * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
}

/** Haversine 대원 거리 (m) */
export function haversineM(lat1: number, lon1: number, lat2: number, lon2: number): number {
  return haversineKm(lat1, lon1, lat2, lon2) * 1000;
}

/** 초기 방위각 (°, 0=N, CW) */
export function bearingDeg(lat1: number, lon1: number, lat2: number, lon2: number): number {
  const dLon = (lon2 - lon1) * DEG2RAD;
  const y = Math.sin(dLon) * Math.cos(lat2 * DEG2RAD);
  const x =
    Math.cos(lat1 * DEG2RAD) * Math.sin(lat2 * DEG2RAD) -
    Math.sin(lat1 * DEG2RAD) * Math.cos(lat2 * DEG2RAD) * Math.cos(dLon);
  let deg = (Math.atan2(y, x) * 180) / Math.PI;
  return (deg + 360) % 360;
}

/** 방위(°) + 거리(km) 통합 계산 (평면 근사, 근거리용) */
export function azimuthAndDist(
  radarLat: number, radarLon: number,
  lat: number, lon: number,
): { azDeg: number; distKm: number } {
  const dLat = lat - radarLat;
  const dLon = lon - radarLon;
  const latKm = dLat * 111.32;
  const lonKm = dLon * 111.32 * Math.cos(radarLat * DEG2RAD);
  const distKm = Math.sqrt(latKm * latKm + lonKm * lonKm);
  let azDeg = (Math.atan2(lonKm, latKm) * 180) / Math.PI;
  if (azDeg < 0) azDeg += 360;
  return { azDeg, distKm };
}
